File: LeetCode/101. Symmetric Tree/recursive_solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if not root:
            return True
        return self.helper(root.left, root.right)
    
    def helper(self, ltree, rtree):
        if not ltree and not rtree:
            return True
        if not ltree or not rtree:
            return False
        if ltree.val == rtree.val:
            return self.helper(ltree.left, rtree.right) and self.helper(ltree.right, rtree.left)
        return False

File: LeetCode/101. Symmetric Tree/recursive_solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        if (root == null) return true;
        return helper(root.left, root.right);
    }
    
    private boolean helper(TreeNode ltree, TreeNode rtree) {
        if (ltree == null && rtree == null) return true;
        if (ltree == null || rtree == null) return false;
        if (ltree.val == rtree.val) {
            return helper(ltree.left, rtree.right) && helper(ltree.right, rtree.left);
        }
        return false;
    }
}
